File: 160. Intersection of Two Linked Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        unordered_map <ListNode*, bool> visit;
        
        while(headA != NULL ) { 
            visit[headA] = true;
            headA = headA->next;
        }
        ListNode* ans = NULL;
        while( headB != NULL ) { 
            if( visit[headB] == true ) { 
                ans = headB;
                break;
            }
            
            headB = headB->next;
        }
        
        return ans;
    }
};

File: javascript/160. Intersection of Two Linked Lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} headA
 * @param {ListNode} headB
 * @return {ListNode}
 */
var getIntersectionNode = function(headA, headB) {
    const cached = new Map();
    while(headA) {
        cached.set(headA, true);
        headA = headA.next;
    }
    let ans = null;
    while( headB ) {
        if( cached.has(headB) ) {
            ans = headB;
            break;
        }
        headB = headB.next;
    }
    return ans;
};
